lib/Dialect/Arc/Transforms/LowerToLLVM.cpp

```diff
@@ -0,0 +1,471 @@
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::OutputOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    rewriter.replaceOpWithNewOp<func::ReturnOp>(op, adaptor.getOutputs());
+    return success();
+  }
+};
+
+struct StateOpLowering : public OpConversionPattern<arc::StateOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::StateOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    SmallVector<Type> newResultTypes;
+    if (failed(
+            typeConverter->convertTypes(op.getResultTypes(), newResultTypes)))
+      return failure();
+    rewriter.replaceOpWithNewOp<func::CallOp>(
+        op, newResultTypes, op.getArcAttr(), adaptor.getInputs());
+    return success();
+  }
+};
+
+struct ClockTreeOpLowering : public OpConversionPattern<arc::ClockTreeOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::ClockTreeOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    if (auto enable = adaptor.getEnable()) {
+      auto ifOp = rewriter.create<scf::IfOp>(op.getLoc(), enable, false);
+      auto yield = ifOp.thenYield();
+      yield->remove();
+      rewriter.mergeBlocks(&op.getBodyBlock(), ifOp.thenBlock(), {});
+      ifOp.thenBlock()->push_back(yield);
+      rewriter.eraseOp(op);
+    } else {
+      Block *parentBlock = op->getBlock();
+      Block *after = rewriter.splitBlock(parentBlock, Block::iterator(op));
+      rewriter.mergeBlocks(&op.getBodyBlock(), parentBlock, {});
+      rewriter.mergeBlocks(after, parentBlock, {});
+      rewriter.eraseOp(op);
+    }
+    return success();
+  }
+};
+
+struct PassThroughOpLowering : public OpConversionPattern<arc::PassThroughOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::PassThroughOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    Block *parentBlock = op->getBlock();
+    Block *after = rewriter.splitBlock(parentBlock, Block::iterator(op));
+    rewriter.mergeBlocks(&op.getBodyBlock(), parentBlock, {});
+    rewriter.mergeBlocks(after, parentBlock, {});
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct AllocStorageOpLowering
+    : public OpConversionPattern<arc::AllocStorageOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AllocStorageOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    auto type = typeConverter->convertType(op.getType());
+    Value offset = rewriter.create<LLVM::ConstantOp>(
+        op.getLoc(), rewriter.getI32Type(), op.getOffsetAttr());
+    rewriter.replaceOpWithNewOp<LLVM::GEPOp>(op, type, adaptor.getInput(),
+                                             offset);
```

Comment:
Awesome, I wasn't aware of that. Looks like it also stores the index in an attribute instead of an operand on the GEP op itself, such that there's no need for a constant at all. Very nice.